File: backend/app/schemas/nearby.py
"""
Lugares próximos (RF-07, RF-08, RF-19, RF-24): os mais perto, com
distância, tempo estimado e nota.
"""

from enum import StrEnum

from app.schemas.base import ApiModel


class NearbyCategory(StrEnum):
    """As 5 categorias da tela (RF-07) e mais duas usadas pelas recomendações."""

    POSTO = "posto"
    RESTAURANTE = "restaurante"
    HOTEL = "hotel"
    PONTO_TURISTICO = "ponto_turistico"
    HOSPITAL = "hospital"
    # DESCANSAR: procura posto ou hotel.
    DESCANSO = "descanso"
    # FAZER UMA PARADA: procura um lugar para pausa.
    PARADA = "parada"


class NearbyPlace(ApiModel):
    id: str
    name: str
    address: str | None = None
    latitude: float
    longitude: float
    # Distância e tempo de carro (calculados pelo OSRM). Se não der para
    # calcular o trajeto, a distância é em linha reta e o tempo fica `null`.
    distance_meters: float
    duration_seconds: float | None = None
    by_road: bool
    # Nota de 1 a 5 e número de avaliações. `null` quando a fonte não tem nota.
    rating: float | None = None
    rating_count: int | None = None


class NearbyResponse(ApiModel):
    category: NearbyCategory
    # De onde vieram os lugares: `google-places`, `tomtom` ou `openstreetmap`.
    source: str
    # Preenchido quando a fonte preferida falhou e foi usada outra.
    fallback_reason: str | None = None
    places: list[NearbyPlace]

File: src/features/nearby/types/nearby.ts
import type { IconName } from '@/components/ui/icon-badge';
import type { ColorToken } from '@/theme/colors';

/**
 * Opções próximas — o formato de `GET /v1/nearby`.
 *
 * As 5 categorias visuais do escopo (RF-07) e duas que só as recomendações
 * usam (§4.7): DESCANSAR busca posto ou hotel, FAZER UMA PARADA busca um local
 * de pausa.
 */
export type NearbyCategory =
  | 'posto'
  | 'restaurante'
  | 'hotel'
  | 'ponto_turistico'
  | 'hospital'
  | 'descanso'
  | 'parada';

export type NearbyPlace = {
  id: string;
  name: string;
  address: string | null;
  latitude: number;
  longitude: number;
  /** De carro quando `byRoad`; senão, em linha reta. */
  distanceMeters: number;
  durationSeconds: number | null;
  byRoad: boolean;
  /** Nota de 1 a 5. `null` quando a fonte não tem nota (TomTom, OpenStreetMap). */
  rating: number | null;
  ratingCount: number | null;
};

export type NearbyResponse = {
  category: NearbyCategory;
  /** `google-places`, `tomtom` ou `openstreetmap`. */
  source: string;
  fallbackReason: string | null;
  places: NearbyPlace[];
};

/** As categorias visuais do escopo, na ordem em que aparecem (RF-07). */
export const SCOPE_CATEGORIES: {
  id: NearbyCategory;
  label: string;
  icon: IconName;
  color: ColorToken;
}[] = [
  { id: 'posto', label: 'Posto', icon: 'gas-station', color: 'categoryFuel' },
  { id: 'restaurante', label: 'Restaurante', icon: 'silverware-fork-knife', color: 'categoryFood' },
  { id: 'hotel', label: 'Hotel', icon: 'bed', color: 'categoryLodging' },
  { id: 'ponto_turistico', label: 'Turismo', icon: 'camera-marker', color: 'categoryNature' },
  { id: 'hospital', label: 'Hospital', icon: 'hospital-box', color: 'categoryHealth' },
];
